File: src/20/day20.py
from collections import namedtuple
from typing import Optional

# type: bool, dests: list[str], state: bool, inputs: Optional[dict[str, bool]]
Module = namedtuple('Module', ['type', 'dests', 'state', 'inputs'])
# NOTE: broadcaster is internally represented as a flip-flop, but treated differently
# True is flip-flop, False is conjunction.
# state: True is high, False is low.
# `inputs` is only valid for conjunction (i.e. dests=False)

# origin: str, dest: str, pulse: bool
Send = namedtuple('Send', ['origin', 'dest', 'pulse'])
# pulse: True is high, False is low.

# This procedure edits `modules` as a mutable reference
# It returns the number of low and high signals sent.
def press_button(modules: dict[str, Module]) -> tuple[int, int]:
    tasks = [Send('broadcaster', d, False) for d in modules['broadcaster'].dests]
    lows, highs = len(tasks) + 1, 0
    while tasks:
        new_tasks = []
        for task in tasks:
            module_name = task.dest
            if modules.get(module_name, None) is None:
                continue
            if modules[module_name].type:
                if not task.pulse:
                    modules[module_name] = Module(
                        modules[module_name].type,
                        modules[module_name].dests,
                        not modules[module_name].state,
                        None
                    )
                    new_pulse = modules[module_name].state
                    for dest in modules[module_name].dests:
                        if new_pulse:
                            highs += 1
                        else:
                            lows += 1
                        new_tasks.append(Send(module_name, dest, new_pulse))
            else:
                modules[module_name].inputs[task.origin] = task.pulse
                new_pulse = any(not signal for signal in modules[module_name].inputs.values())
                for dest in modules[module_name].dests:
                    if new_pulse:
                        highs += 1
                    else:
                        lows += 1
                    new_tasks.append(Send(module_name, dest, new_pulse))
        tasks = new_tasks
    return lows, highs

def part1(filename: str) -> int:
    modules = dict()
    conjunctions = []
    with open(filename, 'rt') as f:
        for line in f:
            if line.startswith('broadcaster'):
                _, dests = line.split(' -> ')
                modules['broadcaster'] = Module(True, dests.strip().split(', '), False, None)
            elif line.startswith('%'):
                name, dests = line.lstrip('%').strip().split(' -> ')
                modules[name] = Module(True, dests.strip().split(', '), False, None)
            elif line.startswith('&'):
                name, dests = line.lstrip('&').strip().split(' -> ')
                modules[name] = Module(False, dests.strip().split(', '), False, dict())
                conjunctions.append(name)
            else:
                break
    for module in modules:
        for dest in modules[module].dests:
            if dest in conjunctions:
                modules[dest].inputs[module] = False

    low_total, high_total = 0, 0
    for _ in range(1000):
        low, high = press_button(modules)
        low_total += low
        high_total += high
    return low_total * high_total

if __name__ == '__main__':
    print(part1('day20_input.txt'))

File: src/20/day20.rs
use std::collections::HashMap;

#[derive(Debug)]
pub struct Module {
    pub module_type: bool,
    pub dests: Vec<String>,
    pub state: bool,
    pub inputs: HashMap<String, bool>
}

#[derive(Debug)]
pub struct Send {
    pub origin: String,
    pub dest: String,
    pub pulse: bool
}

fn press_button(modules: &mut HashMap<String, Module>) -> (u64, u64) {
    let mut tasks = modules
        .get("broadcaster")
        .unwrap()
        .dests
        .iter()
        .map(|d| Send { origin: "broadcaster".to_string(), dest: d.to_string(), pulse: false })
        .collect::<Vec<_>>();
    let mut lows = (tasks.len() + 1) as u64;
    let mut highs = 0u64;
    while tasks.len() > 0 {
        let mut new_tasks = Vec::<Send>::new();
        for task in &tasks {
            let module_name = task.dest.clone();
            if let Some(module) = modules.get(&module_name) {
                if module.module_type {
                    if !task.pulse {
                        modules.insert(
                            module_name.clone(),
                            Module {
                                module_type: module.module_type,
                                dests: module.dests.iter().cloned().collect(),
                                state: !module.state,
                                inputs: HashMap::new()
                            }
                        );
                        let new_pulse = modules.get(&module_name).unwrap().state;
                        for dest in &modules.get(&module_name).unwrap().dests {
                            if new_pulse { highs += 1; } else { lows += 1; }
                            new_tasks.push(
                                Send {
                                    origin: module_name.clone(),
                                    dest: dest.clone(),
                                    pulse: new_pulse
                                }
                            );
                        }
                    }
                } else {
                    modules
                        .get_mut(&module_name)
                        .unwrap()
                        .inputs
                        .insert(task.origin.clone(), task.pulse);
                    let new_pulse = modules
                        .get(&module_name)
                        .unwrap()
                        .inputs
                        .values()
                        .any(|s| !s);
                    for dest in &modules.get(&module_name).unwrap().dests {
                        if new_pulse { highs += 1; } else { lows += 1; }
                        new_tasks.push(
                            Send {
                                origin: module_name.clone(),
                                dest: dest.clone(),
                                pulse: new_pulse
                            }
                        );
                    }
                }
            } else {
                continue;
            }
        }
        tasks = new_tasks;
    }
    return (lows, highs);
}

fn part1(filename: &str) -> u64 {
    let mut modules = HashMap::<String, Module>::new();
    let mut conjunctions = Vec::<String>::new();
    let lines = std::fs::read_to_string(filename).expect("File not found");
    for line in lines.trim().split('\n').map(|x| x.trim()) {
        if line.starts_with("broadcaster") {
            let dests = line.split(" -> ").collect::<Vec<_>>()[1];
            modules.insert(
                "broadcaster".to_string(),
                Module {
                    module_type: true,
                    dests: dests.trim().split(", ").map(|s| s.to_string()).collect::<Vec<_>>(),
                    state: false,
                    inputs: HashMap::<String, bool>::new()
                }
            );
        } else if line.starts_with('%') {
            let mut line_iter = line[1..].trim().split(" -> ");
            let name = line_iter.next().unwrap();
            let dests = line_iter.next().unwrap();
            modules.insert(
                name.to_string(),
                Module {
                    module_type: true,
                    dests: dests.trim().split(", ").map(|s| s.to_string()).collect::<Vec<_>>(),
                    state: false,
                    inputs: HashMap::<String, bool>::new()
                }
            );
        } else if line.starts_with('&') {
            let mut line_iter = line[1..].trim().split(" -> ");
            let name = line_iter.next().unwrap();
            let dests = line_iter.next().unwrap();
            modules.insert(
                name.to_string(),
                Module {
                    module_type: false,
                    dests: dests.trim().split(", ").map(|s| s.to_string()).collect::<Vec<_>>(),
                    state: false,
                    inputs: HashMap::<String, bool>::new()
                }
            );
            conjunctions.push(name.to_string());
        } else {
            break;
        }
    }

    let keys_and_dests = modules
        .iter()
        .map(|(k, v)| (k.clone(), v.dests.iter().cloned().collect::<Vec<_>>()))
        .collect::<Vec<_>>();
    for (module_name, dests) in keys_and_dests {
        for dest in dests {
            if conjunctions.contains(&dest) {
                modules.get_mut(&dest)
                    .unwrap()
                    .inputs
                    .insert(module_name.to_string(), false);
            }
        }
    }
    let mut low_total: u64 = 0;
    let mut high_total: u64 = 0;
    for _ in 0..1000 {
        let (low, high) = press_button(&mut modules);
        low_total += low;
        high_total += high;
    }
    return low_total * high_total;
}

fn main() {
    println!("{}", part1("day20_input.txt"));
}
